### PR TITLE
ConsoleInteraction.py: Change bears list order

### PR DESCRIPTION
**Pull request by corona10**
Bears are now displayed alphabetically case-insensitive.

Closes https://github.com/coala/coala/issues/3941

**For short term contributors:** we understand that getting your commits well
defined like we require is a hard task and takes some learning. If you
look to help without wanting to contribute long term there's no need
for you to learn this. Just drop us a message and we'll take care of brushing
up your stuff for merge!

### Checklist

- [x] I read the [commit guidelines](http://coala.io/commit) and I've followed
      them.
- [x] I ran coala over my code locally. (*All commits have to pass
      individually.* It is not sufficient to have "fixup commits" on your PR,
      our bot will still report the issues for the previous commit.) You will
      likely receive a lot of bot comments and build failures if coala does not
      pass on every single commit!

After you submit your pull request, **DO NOT click the 'Update Branch' button.**
When asked for a rebase, consult [coala.io/rebase](https://coala.io/rebase)
instead.

Please consider helping us by reviewing other peoples pull requests as well:

- pick up any PR at <https://coala.io/review>
- review it (check <https://coala.io/reviewing> for more info)
- if you are sure that it needs work, use `cobot mark wip <URL>` to get it out
  of the review queue.

The more you review, the more your score will grow at coala.io and we will
review your PRs faster!

### CODE REVIEW

File: coalib/output/ConsoleInteraction.py [hunk 1]
```diff
@@ -766,15 +766,18 @@ def print_bears(bears,
         console_printer.print('No bears to show. Did you forget to install '
                               'the `coala-bears` package? Try `pip3 install '
                               'coala-bears`.')
-        return
+        return None
```
Comment (Techievena) [inline coalib/output/ConsoleInteraction.py]: Why this?

File: coalib/output/ConsoleInteraction.py [hunk 2]
```diff
@@ -766,15 +766,18 @@ def print_bears(bears,
...
 
-    for bear, sections in sorted(bears.items(),
-                                 key=lambda bear_tuple: bear_tuple[0].name):
```
Comment (Techievena) [inline coalib/output/ConsoleInteraction.py]: You can just add a `.lower()` here right?

File: coalib/output/ConsoleInteraction.py [hunk 3]
```diff
@@ -796,7 +799,6 @@ def show_bears(local_bears,
     :param console_printer:  Object to print messages on the console.
     """
     bears = inverse_dicts(local_bears, global_bears)
-
```
Comment (Techievena) [inline coalib/output/ConsoleInteraction.py]: This is unnecessary

File: coalib/output/ConsoleInteraction.py (hunk 2 above)
Comment (corona10) [inline coalib/output/ConsoleInteraction.py]: @Techievena Yes, you are right. but I need to add a unit test for it is well printing by case insensitive. So I need to return this. Do you have better idea to check this test?

**Comment by corona10** [comment]
@Techievena  
Okay. So I modified this function for printing by case-insensitive orders.
But this function could be modified in future by someone by messy ordering.
So I want to add a unit test for this function which checks case-insensitive orders.
That is why I add return value for this function because we check it by return value.

**Comment by corona10** [comment]
@Techievena
One way to improve a unit test for this function. Could be capturing stdout for this function.
If this could works, we don't have to modify this function to return something.

**Comment by corona10** [comment]
Okay. I found that `PrintFormattedResultsTest` is already do that. I will using them.

File: coalib/output/ConsoleInteraction.py (hunk 2 above)
Comment (Techievena) [inline coalib/output/ConsoleInteraction.py]: Moreover a function like `print_bears` shouldn't return any value. Not at all a value like `sorted_bears`. Maybe you can test without returning any value from the function.

Comment (Techievena) [inline coalib/output/ConsoleInteraction.py]: Like manipulate the order in which bears are passed to `show_bear`.

**Comment by SanketDG** [comment]
You will not have to return anything, see how `show_bears` is tested https://github.com/coala/coala/blob/master/tests/output/ConsoleInteractionTest.py#L722

File: tests/output/ConsoleInteractionTest.py [hunk 4]
```diff
@@ -11,9 +11,10 @@
 from coalib.bears.Bear import Bear
 from coala_utils.ContextManagers import (
     make_temp, retrieve_stdout, simulate_console_inputs)
+from coalib.misc.DictUtilities import inverse_dicts
 from coalib.output.ConsoleInteraction import (
     acquire_actions_and_apply, acquire_settings, get_action_info, nothing_done,
-    print_affected_files, print_result, print_results,
+    print_affected_files, print_bears, print_result, print_results,
```
Comment (SanketDG) [inline tests/output/ConsoleInteractionTest.py]: you do not need to import `print_bears`, just test `show_bears`

File: tests/output/ConsoleInteractionTest.py [hunk 5]
```diff
@@ -174,6 +175,44 @@ def test_require_settings(self):
 
             self.assertEqual(generator.last_input, 2)
 
+    def test_print_bears(self):
```
Comment (SanketDG) [inline tests/output/ConsoleInteractionTest.py]: Although this is a good way to test your changes, it is not the best.

You should focus on what your change is, that is printing the bears in sorted order, this means to have testcases with both uppercase and lowercase (ideally 3-5) and just print the bear names.

Also please test the `show_bears` function.

File: tests/output/ConsoleInteractionTest.py [hunk 6]
```diff
@@ -174,6 +175,44 @@ def test_require_settings(self):
...
+        bears = inverse_dicts(self.local_bears, self.global_bears)
+        expected = 'No bears to show. Did you forget to install '\
+                   'the `coala-bears` package?'\
+                   ' Try `pip3 install coala-bears`.\n'
+        with retrieve_stdout() as stdout:
+            print_bears({}, True, True, self.console_printer)
```
Comment (SanketDG) [inline tests/output/ConsoleInteractionTest.py]: use `show_bears` here, and pass `False` to the next two params

File: tests/output/ConsoleInteractionTest.py [hunk 7]
```diff
@@ -174,6 +175,44 @@ def test_require_settings(self):
 
             self.assertEqual(generator.last_input, 2)
 
+    def test_print_bears(self):
+        bears = inverse_dicts(self.local_bears, self.global_bears)
+        expected = 'No bears to show. Did you forget to install '\
+                   'the `coala-bears` package?'\
+                   ' Try `pip3 install coala-bears`.\n'
```
Comment (SanketDG) [inline tests/output/ConsoleInteractionTest.py]: this expected string is not required

File: tests/output/ConsoleInteractionTest.py (hunk 6 above)
Comment (SanketDG) [inline tests/output/ConsoleInteractionTest.py]: Also pass a dict to the first argument, see https://github.com/coala/coala/blob/master/tests/output/ConsoleInteractionTest.py#L729-L730

Comment (SanketDG) [inline tests/output/ConsoleInteractionTest.py]: The no of bears i.e. testcases should be ideally less than 5

File: tests/output/ConsoleInteractionTest.py [hunk 8]
```diff
@@ -174,6 +175,44 @@ def test_require_settings(self):
...
+            'it can analyze.\n\n'\
+            '  No needed settings.\n\n''  No optional settings.\n\n'\
+            '  This bear does not provide information about'\
+            ' what categories it can detect.\n\n'\
+            '  This bear cannot fix issues or does not provide'\
+            ' information about what categories it can fix.\n\n'\
+            '  Path:\n'+"   '"+os.path.abspath(__file__)+"'\n\n"
+
+        with retrieve_stdout() as stdout:
+            print_bears(bears, True, True, self.console_printer)
```
Comment (SanketDG) [inline tests/output/ConsoleInteractionTest.py]: this is for https://github.com/coala/coala/pull/3943/files#r106773214

Comment (SanketDG) [inline tests/output/ConsoleInteractionTest.py]: remove the above lines, and use everything here that I said

File: coalib/output/ConsoleInteraction.py (hunk 3 above)
Comment (SanketDG) [inline coalib/output/ConsoleInteraction.py]: remove this newline

**Comment by SanketDG** [comment]
Also please put the tests inside `ShowBearsTest` preferably under `test_show_bears` with name `test_show_bears_sorted`

**Comment by corona10** [comment]
@SanketDG Thanks for the review. I will take a look as your guide!

File: tests/output/ConsoleInteractionTest.py [hunk 9]
```diff
@@ -734,6 +753,25 @@ def test_show_bears(self, show_bear):
                                           True,
                                           self.console_printer)
 
+    def test_show_bears_sorted(self):
+        local_bears = OrderedDict([('default', [SomelocalBear]),
+                                   ('test', [aSomelocalBear])])
+        global_bears = OrderedDict([('default', [SomeglobalBear]),
+                                    ('test', [BSomeglobalBear])])
+        with retrieve_stdout() as stdout:
```
Comment (SanketDG) [inline tests/output/ConsoleInteractionTest.py]: you do not need lines 761 - 766 :smile: they are already tested!

File: tests/output/ConsoleInteractionTest.py [hunk 10]
```diff
@@ -114,6 +115,24 @@ def run(self):
         return None
 
 
+class aSomelocalBear(Bear):
+
+    def run(self):
+        """
+        Some local-bear a Description.
```
Comment (SanketDG) [inline tests/output/ConsoleInteractionTest.py]: Just keep it `Some local-bear Description`, same for below

File: tests/output/ConsoleInteractionTest.py [hunk 11]
```diff
@@ -734,6 +753,25 @@ def test_show_bears(self, show_bear):
...
+            self.assertEqual(stdout.getvalue(
+                ), 'No bears to show. Did you forget to install'
+                   ' the `coala-bears` package?'
+                   ' Try `pip3 install coala-bears`.\n')
+
+        with retrieve_stdout() as stdout:
+            show_bears(local_bears, global_bears, False,
+                       False, self.console_printer)
+            self.assertEqual(stdout.getvalue(
+                ), 'aSomelocalBear\nBSomeglobalBear\n'
```
Comment (SanketDG) [inline tests/output/ConsoleInteractionTest.py]: the wrapping doesn't look right

should be like:

```
self.assertEqual(stdout.getvalue(),
                           'statements here')
```

Comment (SanketDG) [inline tests/output/ConsoleInteractionTest.py]: also could you put the statements in separate lines like:
```
'aSomelocalBear\n'
'BSomeGlobalBear\n'
```

File: tests/output/ConsoleInteractionTest.py [hunk 12]
```diff
@@ -11,6 +11,7 @@
 from coalib.bears.Bear import Bear
 from coala_utils.ContextManagers import (
     make_temp, retrieve_stdout, simulate_console_inputs)
+from coalib.misc.DictUtilities import inverse_dicts
```
Comment (SanketDG) [inline tests/output/ConsoleInteractionTest.py]: unneeded :smile:

**Comment by corona10** [comment]
@SanketDG Thanks. I updated it. PTAL

File: tests/output/ConsoleInteractionTest.py [hunk 13]
```diff
@@ -114,6 +114,24 @@ def run(self):
...
+        Some local-bear Description.
+        """
+        return None
+
+
+class BSomeglobalBear(Bear):
+
+    def run(self):
+        """
+        Some global-bear a Description.
```
Comment (SanketDG) [inline tests/output/ConsoleInteractionTest.py]: you missed out this one

**Comment by corona10** [comment]
@SanketDG PTAL

**Comment by SanketDG** [comment]
The code is okay, the only small nitpick is the commit message.

First is the tag, it should be exact to the filename you are changing

Secondly you will want to give a commit message, that is **complete** and less than 50 chars (I'll leave this upto you 😄 )

The commit body should also describe everything you did, for your case, something like this:

```
Bears are now displayed alphabetically case-insensitive
```

Also "closes" > "Closes"

more at https://coala.io/commit

**Comment by corona10** [comment]
@SanketDG Thanks a lot! I changed as your comment. 👍

**Comment by SanketDG** [comment]
ack c103fed

**Comment by corona10** [comment]
rebase done!

**Comment by SanketDG** [comment]
ack 4e35e5e

**Comment by SanketDG** [comment]
@rultor merge

**Comment by corona10** [comment]
@SanketDG  I rebase it one more time right now.

**Comment by SanketDG** [comment]
oops sorry been a while :smile:
ack 0e437a9

**Comment by SanketDG** [comment]
ack 0e437a9

**Comment by SanketDG** [comment]
@rultor merge

**Pull request merged**